--- /home/test/modes.internal.h_old	2019-07-26 12:59:06.507498171 +0000
+++ src/crypto/fipsmodule/modes/internal.h	2019-07-26 13:00:16.678034966 +0000
@@ -67,11 +67,11 @@
 static inline uint32_t GETU32(const void *in) {
   uint32_t v;
   OPENSSL_memcpy(&v, in, sizeof(v));
-  return CRYPTO_bswap4(v);
+  return CRYPTO_BSWAP4(v);
 }
 
 static inline void PUTU32(void *out, uint32_t v) {
-  v = CRYPTO_bswap4(v);
+  v = CRYPTO_BSWAP4(v);
   OPENSSL_memcpy(out, &v, sizeof(v));
 }
 
