--- /home/test/wnaf.c_old	2019-07-26 11:30:12.736884654 +0000
+++ src/crypto/fipsmodule/ec/wnaf.c	2019-07-26 11:31:06.151989695 +0000
@@ -93,7 +93,7 @@
   int next_bit = bit << 1;  // 2^(w+1), at most 256
   int mask = next_bit - 1;  // at most 255
 
-  int window_val = scalar->words[0] & mask;
+  int window_val = BSWAP_ULONG(scalar->words[0]) & mask;
   for (size_t j = 0; j < bits + 1; j++) {
     assert(0 <= window_val && window_val <= next_bit);
     int digit = 0;
